### PR TITLE
Set language in constructor of \RestfulBase 

### PR DESCRIPTION
**Pull request by Jelle-S**
(follow-up of #318)

### CODE REVIEW

**Comment by e0ipso** [comment]
Awesome PR!

File: plugins/restful/RestfulBase.php [hunk 1]
```diff
@@ -345,7 +371,7 @@ public function getRateLimitManager() {
    * @param DrupalCacheInterface $cache_controller
    *   (optional) Injected cache backend.
    */
-  public function __construct(array $plugin, \RestfulAuthenticationManager $auth_manager = NULL, \DrupalCacheInterface $cache_controller = NULL) {
+  public function __construct(array $plugin, \RestfulAuthenticationManager $auth_manager = NULL, \DrupalCacheInterface $cache_controller = NULL, $langcode = NULL) {
```
Comment (e0ipso) [inline plugins/restful/RestfulBase.php:374]: We will probably need to update the inheriting constructors too.

File: plugins/restful/RestfulBase.php [hunk 2]
```diff
@@ -76,6 +76,13 @@
   protected $valueMetadata = array();
 
   /**
+   * Determines the language of the items that should be returned.
+   *
+   * @var string
+   */
+  protected $langcode;
```
Comment (e0ipso) [inline plugins/restful/RestfulBase.php:83]: I like having it in `\RestfulBase` even that this property is probably only interesting to entities.

**Comment by Jelle-S** [comment]
Inherited constructors are updated ;-)

**Comment by amitaibu** [comment]
Thanks @Jelle-S ! I'm going to push our luck here -- any chance for a simpleTest? :)

File: plugins/restful/RestfulEntityBaseMultipleBundles.php [hunk 3]
```diff
@@ -27,8 +27,8 @@ public static function controllersInfo() {
     );
   }
 
-  public function __construct(array $plugin, \RestfulAuthenticationManager $auth_manager = NULL, \DrupalCacheInterface $cache_controller = NULL) {
-    parent::__construct($plugin);
+  public function __construct(array $plugin, \RestfulAuthenticationManager $auth_manager = NULL, \DrupalCacheInterface $cache_controller = NULL, $language = NULL) {
+    parent::__construct($plugin, $auth_manager, $cache_controller, $language);
```
Comment (e0ipso) [inline plugins/restful/RestfulEntityBaseMultipleBundles.php:31]: Good catch!

**Comment by e0ipso** [comment]
This probably means that we need to adapt our constructor for the Search API integration too. :-)

Awesome work!

**Comment by Jelle-S** [comment]
I'm guessing Travis isn't enabling/downloading Entity Translation? I have no problems on my local machine

**Comment by amitaibu** [comment]
I think we don't need Entity Translation, as we can just enable `translatable` on field. So only need to enable locale.

**Comment by amitaibu** [comment]
> I think we don't need Entity Translation

btw, In case I'm wrong - we can download the module via `.travis.yml`

**Comment by Jelle-S** [comment]
The tests failed withouth Entity Translation (just setting `translatable` wasn't enough)

**Comment by Jelle-S** [comment]
Ah, I didn't apply the two patches for entity on my local setup, updated tests coming soon

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 4]
```diff
@@ -0,0 +1,180 @@
+<?php
+
+/**
+ * @file
+ * Contains RestfulViewEntityMultiLingualTestCase
+ */
+
+class RestfulViewEntityMultiLingualTestCase extends RestfulCurlBaseTestCase {
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test:8]: Thanks for the test!

I think we should trim it down a bit, though. I would expect a smaller test case, where we simply assert a single/ multi value fields for 2 different languages - so the test would be really focused on the functionality, easier to maintain, and easier to follow.

Also, I do believe entity translation shouldn't be needed. For example, the Message module uses only local to provide translatable fields.

**Comment by Jelle-S** [comment]
New (simplified) tests without entity translation

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        'text_processing' => 0,
+      ),
+    );
+    field_create_instance($instance);
+  }
+
+  /**
+   * Test viewing an entity (GET method).
+   *
+   * v1.0 - Simple entity view (id, label, self).
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test]: I believe this can be removed.

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+   * v1.1 - Text and entity reference fields.
+   * v1.2 - "callback" and "process callback".
+   * v1.3 - Non-existing "callback" property.
+   * v1.4 - Non-existing "process callback" property.
+   * v1.6 - XML output format.
+   */
+  function testViewEntity() {
+    $user = $this->drupalCreateUser();
+    $entity = entity_create(
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test]: matter of taste but I would prefer:

``` php
$values = array(
  'name' => 'restful_test_translatable_entity',
);
$entity = entity_create( 'restful_test_translatable_entity', $values);
```

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+      'label' => t('Text multiple'),
+      'settings' => array(
+        'text_processing' => 0,
+      ),
+    );
+    field_create_instance($instance);
+  }
+
+  /**
+   * Test viewing an entity (GET method).
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test]: Maybe better `Test viewing an entity with translatable fields.`

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+   *
+   * v1.0 - Simple entity view (id, label, self).
+   * v1.1 - Text and entity reference fields.
+   * v1.2 - "callback" and "process callback".
+   * v1.3 - Non-existing "callback" property.
+   * v1.4 - Non-existing "process callback" property.
+   * v1.6 - XML output format.
+   */
+  function testViewEntity() {
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test]: `testViewEntity` => `testViewMultiLangualEntity`

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+      array(
+        'name' => 'restful_test_translatable_entity',
+        'uid' => $user->uid,
+        'label' => 'Test translation',
+      )
+    );
+    $wrapper = entity_metadata_wrapper('restful_test_translatable_entity', $entity);
+
+    $text1 = array(
+      'en' => $this->randomName(),
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test:110]: I think out assert should be "smarter" and assert we indeed got the expected results. So we should probably have for example `$text1_en = $this->randomName()`, which we can later compare the results to

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+  /**
+   * Helper to test viewing an entity (GET method) in a certain language.
+   *
+   * @param string $langcode
+   *   The language to view the entity in.
+   * @param int $id
+   *   The ID of the entity to test.
+   */
+  private function _testViewEntityLang($langcode, $id, $text1, $text2) {
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test]: Maybe better: `_testViewEntityLang` => `assertExpectedResult`

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    $handler = restful_get_restful_handler('restful_test_translatable_entity');
+    $handler->setLangCode($langcode);
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test:146]: `// Explicitly set the langcode.`

File: tests/RestfulViewEntityMultiLingualTestCase.test [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    $response = $handler->get($id);
+    $result = $response[0];
+
+    $expected_result = array(
```
Comment (amitaibu) [inline tests/RestfulViewEntityMultiLingualTestCase.test]: We can probably just assert specifically the `text_single`/ `text_multiple` results.

**Comment by amitaibu** [comment]
Thanks @Jelle-S , I've added a few comments. We are getting very close! :)

**Comment by Jelle-S** [comment]
about the `$text1_en`:
That is basically what we do now? we pass `$text1_en[$langcode]` to `_testViewEntityLang()`

**Comment by amitaibu** [comment]
> That is basically what we do now?

Sorry, you are right.

**Comment by Jelle-S** [comment]
The tests have been updated according to your comments.

**Comment by amitaibu** [comment]
Merged, thanks!

**Pull request closed without merging**